In [3]:
#!pip install pandas

  Using cached pandas-2.2.0-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (19 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 61.0/61.0 kB 2.7 MB/s eta 0:00:00
Using cached pandas-2.2.0-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (13.0 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 18.3/18.3 MB 51.5 MB/s eta 0:00:0000:0100:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 345.4/345.4 kB 33.2 MB/s eta 0:00:00


In [13]:
#!pip install requests

In [1]:
import pandas as pd
import numpy as np 

In [2]:
import pandas as pd
data_url = 'https://ourairports.com/countries/US/airports.hxl'
# Load the CSV file into a DataFrame
airport_df = pd.read_csv(data_url)

# Remove the non-useful row (metadata row) if it exists
if airport_df.iloc[0, 0] == "#meta +id":
    airport_df = airport_df.drop(0)

# Save the cleaned DataFrame to a new CSV file named 'all_us_airports.csv'
airport_df.to_csv('all_us_airports.csv', index=False)


In [3]:
airport_data= pd.read_csv('all_us_airports.csv')

In [4]:
airport_data.head()

id ident           type  \
0  3632  KLAX  large_airport   
1  3754  KORD  large_airport   
2  3622  KJFK  large_airport   
3  3384  KATL  large_airport   
4  3878  KSFO  large_airport   

                                               name  latitude_deg  \
0                 Los Angeles International Airport     33.942501   
1              Chicago O'Hare International Airport     41.978600   
2              John F Kennedy International Airport     40.639447   
3  Hartsfield Jackson Atlanta International Airport     33.636700   
4               San Francisco International Airport     37.618999   

   longitude_deg  elevation_ft continent   country_name iso_country  ...  \
0    -118.407997         125.0       NaN  United States          US  ...   
1     -87.904800         672.0       NaN  United States          US  ...   
2     -73.779317          13.0       NaN  United States          US  ...   
3     -84.428101        1026.0       NaN  United States          US  ...   
4    -122.375000          13.0       NaN  United States          US  ...   

    municipality scheduled_service gps_code iata_code  local_code  \
0    Los Angeles                 1     KLAX       LAX         LAX   
1        Chicago                 1     KORD       ORD         ORD   
2       New York                 1     KJFK       JFK         JFK   
3        Atlanta                 1     KATL       ATL         ATL   
4  San Francisco                 1     KSFO       SFO         SFO   

                                           home_link  \
0                            https://www.flylax.com/   
1  https://www.flychicago.com/ohare/home/pages/de...   
2                        https://www.jfkairport.com/   
3                    http://www.atlanta-airport.com/   
4                             http://www.flysfo.com/   

                                      wikipedia_link  \
0  https://en.wikipedia.org/wiki/Los_Angeles_Inte...   
1  https://en.wikipedia.org/wiki/O'Hare_Internati...   
2  https://en.wikipedia.org/wiki/John_F._Kennedy_...   
3  https://en.wikipedia.org/wiki/Hartsfield–Jacks...   
4  https://en.wikipedia.org/wiki/San_Francisco_In...   

                                            keywords    score  \
0                                        Tom Bradley  1335475   
1                                 CHI, Orchard Place  1503175   
2  Manhattan, New York City, NYC, Idlewild, IDL, ...  1052075   
3                                                NaN  2002475   
4                                           QSF, QBA  1112475   

                last_updated  
0  2023-12-21T12:31:02+00:00  
1  2018-09-16T02:35:35+00:00  
2  2022-10-18T18:49:55+00:00  
3  2018-09-19T14:50:01+00:00  
4  2008-06-13T14:30:04+00:00  

[5 rows x 23 columns]

In [5]:
airport_data.shape

(30916, 23)

In [6]:
airport_data.isnull().sum()

id                       0
ident                    0
type                     0
name                     0
latitude_deg             0
longitude_deg            0
elevation_ft          2004
continent            30806
country_name             0
iso_country              0
region_name              0
iso_region               0
local_region             0
municipality            58
scheduled_service        0
gps_code             10614
iata_code            28945
local_code            9689
home_link            29757
wikipedia_link       28327
keywords             24772
score                    0
last_updated             0
dtype: int64

In [7]:
airport_data.dropna(subset=['iata_code'], inplace=True)

In [8]:
airport_data.shape

(1971, 23)

In [9]:
airport_data.isnull().sum()

id                      0
ident                   0
type                    0
name                    0
latitude_deg            0
longitude_deg           0
elevation_ft           30
continent            1952
country_name            0
iso_country             0
region_name             0
iso_region              0
local_region            0
municipality            0
scheduled_service       0
gps_code               75
iata_code               0
local_code             14
home_link            1192
wikipedia_link        257
keywords             1488
score                   0
last_updated            0
dtype: int64

In [10]:

large_airports = airport_data[airport_data['type'] == 'large_airport']


relevant_columns = ['name', 'latitude_deg', 'longitude_deg', 'iata_code']
large_airports_filtered = large_airports[relevant_columns]


large_airports_filtered.to_csv('large_us_airports.csv', index=False)


num_large_airports = len(large_airports_filtered)
print("Number of large airports:", num_large_airports)


Number of large airports: 66


In [11]:
!pip install folium


  Using cached folium-0.15.1-py2.py3-none-any.whl.metadata (3.4 kB)
  Using cached branca-0.7.1-py3-none-any.whl.metadata (1.5 kB)
  Using cached Jinja2-3.1.3-py3-none-any.whl.metadata (3.3 kB)
  Using cached xyzservices-2023.10.1-py3-none-any.whl.metadata (4.0 kB)
  Using cached MarkupSafe-2.1.5-cp38-cp38-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (3.0 kB)
Using cached folium-0.15.1-py2.py3-none-any.whl (97 kB)
Using cached branca-0.7.1-py3-none-any.whl (25 kB)
Using cached Jinja2-3.1.3-py3-none-any.whl (133 kB)
Using cached xyzservices-2023.10.1-py3-none-any.whl (56 kB)
Using cached MarkupSafe-2.1.5-cp38-cp38-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (26 kB)


In [12]:
import folium

# Load the filtered DataFrame
large_airports = pd.read_csv('large_us_airports.csv')

# Create a map centered at an initial location
map_center = [large_airports['latitude_deg'].mean(), large_airports['longitude_deg'].mean()]
mymap = folium.Map(location=map_center, zoom_start=5)

# Add markers for each large airport
for index, row in large_airports.iterrows():
    # Define popup content with airport name and IATA code
    popup_content = f"{row['name']} ({row['iata_code']})"

    # Add marker with custom icon and popup
    folium.Marker(location=[row['latitude_deg'], row['longitude_deg']],
                  popup=popup_content,
                  icon=folium.Icon(icon='plane', prefix='fa')).add_to(mymap)

# Save the map as an HTML file
mymap.save('large_airports_map.html')


In [79]:
import folium


large_airports = pd.read_csv('large_us_airports.csv')


map_center = [large_airports['latitude_deg'].mean(), large_airports['longitude_deg'].mean()]
mymap = folium.Map(location=map_center, zoom_start=5)


for index, row in large_airports.iterrows():
    
    popup_content = f"{row['name']} ({row['iata_code']})"

    
    folium.Marker(location=[row['latitude_deg'], row['longitude_deg']],
                  popup=popup_content,
                  icon=folium.Icon(icon='plane', prefix='fa',icon_rotation=45)).add_to(mymap)


mymap.save('large_airports_map1.html')

next 

In [75]:
import requests

url = "https://api.openaq.org/v2/measurements?date_from=2023-06-06T00:00:00z%3A00%3A00Z&date_to=2023-06-06T13%3A54%3A00Z&limit=100000&sort=desc&radius=7500&order_by=datetime&parameter_id=2&coordinates=42.33143000,-83.0457500"

headers = {"accept": "application/json"}

response = requests.get(url, headers=headers)



In [29]:
import requests
date_from = "2023-06-06"
date_to  = "2023-06-06"
limit=10000
radius = 7500
parameter_id=2
coordinates="42.33143000,-83.04575000"


url = f"https://api.openaq.org/v2/measurements?date_from="+date_from+"T00%3A00%3A00Z&date_to="+date_to+f"T23%3A59%3A59Z"+f"&limit={limit}&sort=desc&radius={radius}&order_by=datetime&parameter_id={parameter_id}&coordinates="+coordinates
headers = {"accept": "application/json"}

response = requests.get(url, headers=headers)

In [33]:
main_data=pd.DataFrame(results)

In [34]:
main_data.head()

locationId          location parameter  value  \
0        1385  Windsor Downtown      pm25  125.0   
1        1385  Windsor Downtown      pm25   21.0   
2        1385  Windsor Downtown      pm25   14.0   
3        1385  Windsor Downtown      pm25   18.0   
4        1385  Windsor Downtown      pm25   21.0   

                                                date   unit  \
0  {'utc': '2023-06-06T23:00:00+00:00', 'local': ...  µg/m³   
1  {'utc': '2023-06-06T22:00:00+00:00', 'local': ...  µg/m³   
2  {'utc': '2023-06-06T21:00:00+00:00', 'local': ...  µg/m³   
3  {'utc': '2023-06-06T20:00:00+00:00', 'local': ...  µg/m³   
4  {'utc': '2023-06-06T19:00:00+00:00', 'local': ...  µg/m³   

                                    coordinates country  city  isMobile  \
0  {'latitude': 42.3144, 'longitude': -83.0433}      US  None     False   
1  {'latitude': 42.3144, 'longitude': -83.0433}      US  None     False   
2  {'latitude': 42.3144, 'longitude': -83.0433}      US  None     False   
3  {'latitude': 42.3144, 'longitude': -83.0433}      US  None     False   
4  {'latitude': 42.3144, 'longitude': -83.0433}      US  None     False   

  isAnalysis                     entity       sensorType  
0       None  Governmental Organization  reference grade  
1       None  Governmental Organization  reference grade  
2       None  Governmental Organization  reference grade  
3       None  Governmental Organization  reference grade  
4       None  Governmental Organization  reference grade

In [43]:

# Transform 'coordinates' into 'sensor_lat' and 'sensor_lon'
main_data['sensor_lat'] = main_data['coordinates'].apply(lambda x: x['latitude'])
main_data['sensor_lon'] = main_data['coordinates'].apply(lambda x: x['longitude'])

# Convert 'date' to 'local_time' in datetime64 format
main_data['local_time'] = pd.to_datetime(main_data['date'].apply(lambda x: x['local']))

# Include only specified fields and filter by 'entity'
filtered_df = main_data[main_data['entity'].isin(['Governmental Organization', 'Community Organization'])]
final_df1 = filtered_df[['locationId', 'location', 'entity', 'parameter', 'value', 'sensor_lat', 'sensor_lon', 'local_time']]


In [44]:
final_df1.head()

locationId          location                     entity parameter  value  \
0        1385  Windsor Downtown  Governmental Organization      pm25  125.0   
1        1385  Windsor Downtown  Governmental Organization      pm25   21.0   
2        1385  Windsor Downtown  Governmental Organization      pm25   14.0   
3        1385  Windsor Downtown  Governmental Organization      pm25   18.0   
4        1385  Windsor Downtown  Governmental Organization      pm25   21.0   

   sensor_lat  sensor_lon                local_time  
0     42.3144    -83.0433 2023-06-06 19:00:00-04:00  
1     42.3144    -83.0433 2023-06-06 18:00:00-04:00  
2     42.3144    -83.0433 2023-06-06 17:00:00-04:00  
3     42.3144    -83.0433 2023-06-06 16:00:00-04:00  
4     42.3144    -83.0433 2023-06-06 15:00:00-04:00

In [45]:
final_df1.shape

(6284, 8)

In [46]:
final_df1.to_csv('20230606_detroit_downtown_7_5km_aq.csv')

In [48]:
import pandas as pd

# Assuming 'final_df1' is your DataFrame loaded from the CSV file
# For demonstration, let's pretend 'final_df1' is already defined with the necessary data

# 1. Mean and median PM2.5 reading over all sensors
mean_pm25 = final_df1['value'].mean()
median_pm25 = final_df1['value'].median()
print(f"Mean PM2.5 reading: {mean_pm25}")
print(f"Median PM2.5 reading: {median_pm25}")

# 2. Standard deviation of PM2.5 readings
std_dev_pm25 = final_df1['value'].std()
print(f"Standard deviation of PM2.5 readings: {std_dev_pm25}")

# 3. Location_id recording the highest PM2.5 and its reading
highest_pm25 = final_df1.loc[final_df1['value'].idxmax(), ['locationId', 'value']]
print(f"Location ID with the highest PM2.5 reading: {highest_pm25['locationId']}")
print(f"Highest PM2.5 reading: {highest_pm25['value']}")

# 4. Ratio of Community Organization to Governmental Organization entity type
entity_types = final_df1['entity'].value_counts()
ratio_community_to_gov = entity_types['Community Organization'] / entity_types['Governmental Organization']
print(f"Ratio of Community Organization to Governmental Organization: {ratio_community_to_gov}")

# 5. Number of unique sensor stations
unique_sensors = final_df1['locationId'].nunique()
print(f"Number of unique sensor stations: {unique_sensors}")

# 6. Station density per km (assuming the area is 7.5 km² as mentioned)
area_km2 = 7.5  # This is an assumption; replace with actual area if known
station_density = unique_sensors / area_km2
print(f"Station density per km²: {station_density}")

# 7. Daily statistics separately for each entity type
# Convert 'local_time' to datetime if it's not already
final_df1['local_time'] = pd.to_datetime(final_df1['local_time'])

# Group by entity and date, then calculate statistics
daily_stats = final_df1.groupby(['entity', final_df1['local_time'].dt.date]).agg(
    mean_pm25=('value', 'mean'),
    median_pm25=('value', 'median'),
    min_pm25=('value', 'min'),
    max_pm25=('value', 'max'),
    std_pm25=('value', 'std'),
    percentile_75=('value', lambda x: x.quantile(0.75))
).reset_index()

# Print daily statistics for each entity type
print("Daily statistics for each entity type:")
print(daily_stats)



Mean PM2.5 reading: 66.0975970719287
Median PM2.5 reading: 37.05
Standard deviation of PM2.5 readings: 216.46080063153576
Location ID with the highest PM2.5 reading: 65949
Highest PM2.5 reading: 1759.5
Ratio of Community Organization to Governmental Organization: 67.30434782608695
Number of unique sensor stations: 15
Station density per km²: 2.0
Daily statistics for each entity type:
                      entity  local_time  mean_pm25  median_pm25  min_pm25  \
0     Community Organization  2023-06-05  34.081768        35.70      21.3   
1     Community Organization  2023-06-06  73.460674        38.50       6.1   
2  Governmental Organization  2023-06-05  28.458333        28.55      26.4   
3  Governmental Organization  2023-06-06  36.410000        35.75      14.0   

   max_pm25    std_pm25  percentile_75  
0      49.6    5.645283         37.700  
1    1759.5  239.515061         46.500  
2      30.3    1.244960         29.200  
3     125.0   15.392811         46.825  


/tmp/ipykernel_105/4178701715.py:37: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  final_df1['local_time'] = pd.to_datetime(final_df1['local_time'])


### PM2.5 Reading Analysis:
- **Mean PM2.5 across all sensors**: 66.1 µg/m³, indicating significant pollution levels. The **median** (37.05 µg/m³) suggests a right-skewed distribution, likely due to outliers.
- **Standard deviation**: 216.46 µg/m³, highlighting a wide range of PM2.5 readings.
- **Highest PM2.5 reading**: 1759.5 µg/m³ at location ID 65949, an extreme pollution event.
- **Ratio of Community to Governmental Organization**: Approximately 67:1, suggesting a greater number of Community Organization readings.
- **Unique sensor stations**: 15, with a **station density** of 2.0 stations per km².

### Daily Statistics (June 5th and 6th):
- Community Organization sensors show a drastic increase in mean PM2.5 from 34.08 to 73.46 µg/m³, influenced by the extreme reading of 1759.5 µg/m³.
- Governmental Organization sensors exhibit more stable mean readings, with less variation between days.
- The **75th percentile** for both entity types increases from June 5th to June 6th, indicating a general rise in PM2.5 levels, aside from the extreme outlier.

### Conclusion:
The mean PM2.5 is significantly influenced by outliers, especially evident in Community Organization readings. The 75th percentile provides a clearer picture of typical air quality, indicating an increase in PM2.5 levels from June 5th to June 6th. The high ratio of Community to Governmental readings suggests extensive monitoring coverage by community sensors.


In [54]:
import folium

# Assuming 'df' is your DataFrame containing the sensor data

# Create a base map of Detroit at a certain zoom level
detroit_map = folium.Map(location=[42.3314, -83.0458], zoom_start=12)

# Add a pin for downtown Detroit
folium.Marker(
    [42.3314, -83.0458], # Coordinates for downtown Detroit
    popup='Downtown Detroit',
    icon=folium.Icon(color='blue', icon='city', prefix='fa')
).add_to(detroit_map)

# Loop through the DataFrame to add community and government sensors
for index, row in final_df1.iterrows():
    if row['entity'] == 'Community Organization':
        # Add a green pin for community sensors
        folium.Marker(
            [row['sensor_lat'], row['sensor_lon']],
            popup=f"{row['location']} (Community)",
            icon=folium.Icon(color='green', icon='info-sign', prefix='glyphicon')
        ).add_to(detroit_map)
    elif row['entity'] == 'Governmental Organization':
        # Add a red pin for government sensors
        folium.Marker(
            [row['sensor_lat'], row['sensor_lon']],
            popup=f"{row['location']} (Government)",
            icon=folium.Icon(color='red', icon='asterisk', prefix='glyphicon')
        ).add_to(detroit_map)

# Display the map
detroit_map
detroit_map.save('detroitmap.html')

